[core] Fix NativeEventEmitter warning on android with jsc

Kudo opened the pull request:
# Why

fix the `WARN  `new NativeEventEmitter()` was called with a non-null argument without the required `addListener` method.` warning on jsc

# How

#19704 doesn't work on jsc which `Object.defineProperties` will add the property to `getPropertyNames`, but jsc will still [call host object getter which returns undefined](https://github.com/expo/expo/blob/c83d5c36787cf93c01a72fa5809dd72b6c202516/packages/expo-modules-core/android/src/main/cpp/JavaScriptModuleObject.cpp#L173). in summary, we cannot use `Object.defineProperties` to override host object properties on jsc.

this pr tries to add the `addListener` and `removeListeners` from the host object.

# Test Plan

- ✅ unit test `JavaScriptModuleObjectTest.hostObject_should_expose_addListener_when_events_dsl_exist`
- ✅ unit test `JavaScriptModuleObjectTest.hostObject_should_not_expose_addListener_by_default`
- ✅ bare-expo + NCL clipboard + hermes
- ✅ bare-expo + NCL clipboard + jsc
- ✅ https://github.com/brentvatne/settings-example


# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

Kudo committed:
[core] Fix NativeEventEmitter warning on android with jsc

Kudo committed:
update changelog

Kudo committed:
add test

lukmccall reviewed (commented):
Looking good, but I think we can do it a little bit better without the CPP part.

What do you think about adding:
```kt
init {
  Function("addListener") { }
  Function("removeListeners") { }
}
```
to the `ObjectDefinitionBuilder`.  

That snippet will add those functions automatically to all module definitions. Leading to the same result without any additional logic in the CPP.

Kudo committed:
Apply suggestions from code review

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/objects/ObjectDefinitionBuilder.kt:
```diff
@@ -253,6 +253,12 @@ open class ObjectDefinitionBuilder {
+    // Register stub functions to bypass react-native `NativeEventEmitter` warnings
+    // WARN  `new NativeEventEmitter()` was called with a non-null argument without the required `addListener` method.
+    // WARN  `new NativeEventEmitter()` was called with a non-null argument without the required `removeListeners` method.
+    Function("addListener") { }
+    Function("removeListeners") { }
```
We could in theory check if those functions weren't added by the user, but I'm not sure if that is needed. What do you think?

Kudo replied:
nice suggestion! https://github.com/expo/expo/pull/19920/commits/7ab54f02e165b0d91c33ff555f2583237437fc2d updates this. does the code change make sense to you?

Kudo committed:
Fix convertJSIArgsToJNI crash

the two function signatures in react-native are:

```js
addListener: (eventName: string) => void;
removeListeners: (count: number) => void;
```

original solution will crash in `MethodMetadata::convertJSIArgsToJNI`
where the expected `argTypes` is empty array but react-native passing a
string or integer.

Kudo committed:
should not override existing functions
